### PR TITLE
feat: add fail safe intervals to restart watches as necessary

### PR DESCRIPTION
**Pull request by alewitt2**

### CODE REVIEW

**Comment by alewitt2** [comment]
https://github.com/razee-io/Razee/issues/135

File: lib/EventHandler.js [hunk 1]
```diff
@@ -31,6 +32,15 @@ module.exports = class EventHandler {
         this._livenessInterval = 60000; //One minute
       }
     }
+    this._livenessEndHours = params.restartPod; // time in hours when we should stop updating the liveness file, which should tell kube to restart our pod. If 0 is defined, pod will not be told to restart. Default: 0.
+    if (!Number.isInteger(this._livenessEndHours)) {
+      this._livenessEndHours = (Number.parseInt(this._livenessEndHours) >= 0) ? Number.parseInt(this._livenessEndHours) : 0;
+    }
```
Comment (adamkingit) [inline lib/EventHandler.js:38]:
```suggestion
    if (!Number.isInteger(this._livenessEndHours)) {
      const livenessEndHours = Number.parseInt(this._livenessEndHours);
      this._livenessEndHours = (livenessEndHours >= 0) ? livenessEndHours : 0;
    }
```

File: lib/EventHandler.js [hunk 2]
```diff
@@ -31,6 +32,15 @@ module.exports = class EventHandler {
...
+    this._watchTimeoutSeconds = objectPath.get(params, 'requestOptions.qs.timeoutSeconds', 300); // time in seconds when the watch should be re-initialized. default: 300s
+    if (!Number.isInteger(this._watchTimeoutSeconds)) {
+      this._watchTimeoutSeconds = Number.parseInt(this._watchTimeoutSeconds) > 0 ? Number.parseInt(this._watchTimeoutSeconds) : 300;
+      this._watchTimeoutMilliseconds = this._watchTimeoutSeconds * 1000;
+    }
```
Comment (adamkingit) [inline lib/EventHandler.js:43]:
```suggestion
    if (!Number.isInteger(this._watchTimeoutSeconds)) {
      const watchTimeoutSeconds = Number.parseInt(this._watchTimeoutSeconds)
      this._watchTimeoutSeconds =  ( watchTimeoutSeconds > 0) ? watchTimeoutSeconds : 300;
      this._watchTimeoutMilliseconds = this._watchTimeoutSeconds * 1000;
    }
```

File: lib/EventHandler.js [hunk 3]
```diff
@@ -55,14 +65,33 @@ module.exports = class EventHandler {
...
+
     if (this._livenessInterval) {
       touch.sync('/tmp/liveness');
+      let livenessStartTime = Date.now(); // time liveness was first initialized
       setInterval(() => {
-        if (this._wm.watching) {
+        let now = Date.now();
+        let millisecondsSinceLivenessStart = now - livenessStartTime;
+        let hoursSinceLivenessStart = millisecondsSinceLivenessStart / 1000 / 60 / 60;
+        let restartThresholdHours = this._livenessEndHours; // stop touching liveness file after the specified number of hours, which should tell kube to restart the pod
```
Comment (adamkingit) [inline lib/EventHandler.js:76]:
```suggestion
        const now = Date.now();
        const millisecondsSinceLivenessStart = now - livenessStartTime;
        const hoursSinceLivenessStart = millisecondsSinceLivenessStart / 1000 / 60 / 60;
        const restartThresholdHours = this._livenessEndHours; // stop touching liveness file after the specified number of hours, which should tell kube to restart the pod
```

File: lib/EventHandler.js [hunk 4]
```diff
@@ -55,14 +65,33 @@ module.exports = class EventHandler {
...
           touch.sync('/tmp/liveness');
         }
       }, this._livenessInterval);
     }
+
+    // every cycle, check if the watch is restarting in the appropriate amount of time
+    setInterval(() => {
+      let now = Date.now();
+      let millisecondsSinceLastStart = now - this._wm.watchStart;
+      let resetThresholdMilliseconds = (this._watchTimeoutMilliseconds + 10000); // interval time and kube timing is not exact, so adding 10 seconds to allow for variance without having to wait for an entire extra cycle
```
Comment (adamkingit) [inline lib/EventHandler.js:87]:
```suggestion
      const now = Date.now();
      const millisecondsSinceLastStart = now - this._wm.watchStart;
      const resetThresholdMilliseconds = (this._watchTimeoutMilliseconds + 10000); // interval time and kube timing is not exact, so adding 10 seconds to allow for variance without having to wait for an entire extra cycle
```

**Review by adamkingit — APPROVED** [review]

**Review by adamkingit — APPROVED** [review]

**Pull request merged**